<a href="https://colab.research.google.com/github/Khalil-Salama/Car_insurance-_claim/blob/main/Insurance_car_claim.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

- Stakeholder
 - Insurance company
- Task
 - To Know them which feature produces the best accuracy for predicting whether a customer will make a car insurance claim. based on the data they provided us with.

Features dictionary

- id: Unique client identifier

- age: Client's age

- gender:	Client's gender

- driving experience: Years the client has been driving:

- education:	Client's level of education:

- income:	Client's income level

- credit_score:	Client's credit score (between zero and one)

- vehicle_ownership	Client's vehicle ownership status

 - 0: Does not own their vehilce (paying off finance)

 - 1: Owns their vehicle

- vehcile_year:	Year of vehicle registration:

 - 0: Before 2015
 - 1: 2015 or later

- married:	Client's marital status:
 - 0: Not married
 - 1: Married

- children:	Client's number of children

- postal_code:	Client's postal code

- annual_mileage:	Number of miles driven by the client each year
- vehicle_type	Type of car:
 - 0: Sedan
 - 1: Sports car

- speeding_violations:	Total number of speeding violations received by the client

- duis:	Number of times the client has been caught driving under the influence of alcohol

- past_accidents:	Total number of previous accidents the client has been involved in

- outcome:	Whether the client made a claim on their car insurance (response variable):

 - 0: No claim
 - 1: Made a claim

In [24]:
###mount google drive
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [25]:
###imports
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


In [26]:
fpath='/content/drive/MyDrive/Colab Notebooks/AXSOSACADEMY/05-IntermediateML/Car_Insurance_Claim.csv'
df=pd.read_csv(fpath)
df.head()

ID    AGE  GENDER      RACE DRIVING_EXPERIENCE    EDUCATION  \
0  569520    65+  female  majority               0-9y  high school   
1  750365  16-25    male  majority               0-9y         none   
2  199901  16-25  female  majority               0-9y  high school   
3  478866  16-25    male  majority               0-9y   university   
4  731664  26-39    male  majority             10-19y         none   

          INCOME  CREDIT_SCORE  VEHICLE_OWNERSHIP VEHICLE_YEAR  MARRIED  \
0    upper class      0.629027                1.0   after 2015      0.0   
1        poverty      0.357757                0.0  before 2015      0.0   
2  working class      0.493146                1.0  before 2015      0.0   
3  working class      0.206013                1.0  before 2015      0.0   
4  working class      0.388366                1.0  before 2015      0.0   

   CHILDREN  POSTAL_CODE  ANNUAL_MILEAGE VEHICLE_TYPE  SPEEDING_VIOLATIONS  \
0       1.0        10238         12000.0        sedan                    0   
1       0.0        10238         16000.0        sedan                    0   
2       0.0        10238         11000.0        sedan                    0   
3       1.0        32765         11000.0        sedan                    0   
4       0.0        32765         12000.0        sedan                    2   

   DUIS  PAST_ACCIDENTS  OUTCOME  
0     0               0      0.0  
1     0               0      1.0  
2     0               0      0.0  
3     0               0      0.0  
4     0               1      1.0

In [27]:
###Getting data frame summery
def summarize_df(df_):
    df = df_.copy()
    report = pd.DataFrame({'dtype':df.dtypes,'null counts': df.isna().sum(),'null(%)': df.isna().sum()/len(df)*100,
                        'nunique':df.nunique(),"min":df.min(),'max':df.max()})
    report.index.name='Column'
    return report.reset_index()
display(summarize_df(df))
display((df.info()))

Column    dtype  null counts  null(%)  nunique           min  \
0                    ID    int64            0     0.00    10000           101   
1                   AGE   object            0     0.00        4         16-25   
2                GENDER   object            0     0.00        2        female   
3                  RACE   object            0     0.00        2      majority   
4    DRIVING_EXPERIENCE   object            0     0.00        4          0-9y   
5             EDUCATION   object            0     0.00        3   high school   
6                INCOME   object            0     0.00        4  middle class   
7          CREDIT_SCORE  float64          982     9.82     9018      0.053358   
8     VEHICLE_OWNERSHIP  float64            0     0.00        2           0.0   
9          VEHICLE_YEAR   object            0     0.00        2    after 2015   
10              MARRIED  float64            0     0.00        2           0.0   
11             CHILDREN  float64            0     0.00        2           0.0   
12          POSTAL_CODE    int64            0     0.00        4         10238   
13       ANNUAL_MILEAGE  float64          957     9.57       21        2000.0   
14         VEHICLE_TYPE   object            0     0.00        2         sedan   
15  SPEEDING_VIOLATIONS    int64            0     0.00       21             0   
16                 DUIS    int64            0     0.00        7             0   
17       PAST_ACCIDENTS    int64            0     0.00       15             0   
18              OUTCOME  float64            0     0.00        2           0.0   

              max  
0          999976  
1             65+  
2            male  
3        minority  
4            30y+  
5      university  
6   working class  
7        0.960819  
8             1.0  
9     before 2015  
10            1.0  
11            1.0  
12          92101  
13        22000.0  
14     sports car  
15             22  
16              6  
17             15  
18            1.0

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 19 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   10000 non-null  int64  
 1   AGE                  10000 non-null  object 
 2   GENDER               10000 non-null  object 
 3   RACE                 10000 non-null  object 
 4   DRIVING_EXPERIENCE   10000 non-null  object 
 5   EDUCATION            10000 non-null  object 
 6   INCOME               10000 non-null  object 
 7   CREDIT_SCORE         9018 non-null   float64
 8   VEHICLE_OWNERSHIP    10000 non-null  float64
 9   VEHICLE_YEAR         10000 non-null  object 
 10  MARRIED              10000 non-null  float64
 11  CHILDREN             10000 non-null  float64
 12  POSTAL_CODE          10000 non-null  int64  
 13  ANNUAL_MILEAGE       9043 non-null   float64
 14  VEHICLE_TYPE         10000 non-null  object 
 15  SPEEDING_VIOLATIONS  10000 non-null  

None

- We have 10,000 rows
- We have 19 columns 11 numeric columns and 8 object columns
- We have Null values in Credit score and in Annual milleage

### Checking for duplicates

In [28]:
###Checking for duplicates
df.duplicated().sum()

np.int64(0)

-No Duplicates were found

### Dropping column ID because it represents the ID number of each indvisual 10000 nunique


In [29]:
###Dropping ID feature
df.drop(columns=['ID'],inplace=True)

### Dropping column Credit score it represents the credit score for each indivisual 9018 nunique and the rest is Nulls.

In [30]:
###Dropping column credit score
df.drop(columns=['CREDIT_SCORE'],inplace=True)

### Droping column postal code because it dosent give us any usefull information about our analysis.

In [31]:
###dropping postal code
df.drop(columns=['POSTAL_CODE'],inplace=True)

### Checking Categorial inconsistancies

In [32]:
##Cat_cols_filter
Cat_cols=df.select_dtypes(include='object').columns
Cat_cols

Index(['AGE', 'GENDER', 'RACE', 'DRIVING_EXPERIENCE', 'EDUCATION', 'INCOME',
       'VEHICLE_YEAR', 'VEHICLE_TYPE'],
      dtype='object')

In [33]:
###loop to check value counts for cat_cols
for col in Cat_cols:
  print(df[col].value_counts())

AGE
26-39    3063
40-64    2931
16-25    2016
65+      1990
Name: count, dtype: int64
GENDER
female    5010
male      4990
Name: count, dtype: int64
RACE
majority    9012
minority     988
Name: count, dtype: int64
DRIVING_EXPERIENCE
0-9y      3530
10-19y    3299
20-29y    2119
30y+      1052
Name: count, dtype: int64
EDUCATION
high school    4157
university     3928
none           1915
Name: count, dtype: int64
INCOME
upper class      4336
middle class     2138
poverty          1814
working class    1712
Name: count, dtype: int64
VEHICLE_YEAR
before 2015    6967
after 2015     3033
Name: count, dtype: int64
VEHICLE_TYPE
sedan         9523
sports car     477
Name: count, dtype: int64


- All categorical features have no inconsistancies and nothing needs to be converted to numerical type
- No null values in categorical features

### Numerical features

In [34]:
###num col fillter
num_cols=df.select_dtypes(include='number').columns
num_cols

Index(['VEHICLE_OWNERSHIP', 'MARRIED', 'CHILDREN', 'ANNUAL_MILEAGE',
       'SPEEDING_VIOLATIONS', 'DUIS', 'PAST_ACCIDENTS', 'OUTCOME'],
      dtype='object')

In [35]:
###Making a loop to describe num cols
for col in num_cols:
  print(df[col].describe())
  print()

count    10000.000000
mean         0.697000
std          0.459578
min          0.000000
25%          0.000000
50%          1.000000
75%          1.000000
max          1.000000
Name: VEHICLE_OWNERSHIP, dtype: float64

count    10000.000000
mean         0.498200
std          0.500022
min          0.000000
25%          0.000000
50%          0.000000
75%          1.000000
max          1.000000
Name: MARRIED, dtype: float64

count    10000.000000
mean         0.688800
std          0.463008
min          0.000000
25%          0.000000
50%          1.000000
75%          1.000000
max          1.000000
Name: CHILDREN, dtype: float64

count     9043.000000
mean     11697.003207
std       2818.434528
min       2000.000000
25%      10000.000000
50%      12000.000000
75%      14000.000000
max      22000.000000
Name: ANNUAL_MILEAGE, dtype: float64

count    10000.000000
mean         1.482900
std          2.241966
min          0.000000
25%          0.000000
50%          0.000000
75%          2.000000
